Add DKG Observer

jonas-lj opened the pull request.

jonas-lj committed:
Add DKG Observer

jonas-lj committed:
Add convenience method

akichidis commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +985,488 @@ where
+pub struct ObserverOutput<G: GroupElement, EG: GroupElement> {
+    pub nodes: Nodes<EG>,
+    pub vss_pk: Poly<G>,
+}
```
Currently `Output` bears the following structure

```
pub struct Output<G: GroupElement, EG: GroupElement> {
    pub nodes: Nodes<EG>,
    pub vss_pk: Poly<G>,
    pub shares: Option<Vec<Share<G::ScalarType>>>, // None if some shares are missing or weight is zero.
}
```

where the `shares` is already an `Option`. Can the Observer use the `Output` as well in this case and just pass the `shares` as `None`? It would help on the API alignment between `Party` and `Observer`

akichidis commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +985,488 @@ where
+    }
+
+    /// Execute the previous two steps together.
+    pub fn complete<R: AllowedRng>(
```
Why is the generic `R: AllowedRng` needed?

jonas-lj replied:
You're right - it's not used anymore. Probably leftovers from the code I copied.

jonas-lj committed:
Use Output for Observer to align Party and Observer APIs

Removes the dedicated ObserverOutput struct and reuses Output<G, EG>
with shares set to None, per review feedback on PR #901. Also fixes
pre-existing test compile errors that called process_message_and_check_pk
on Party (the method exists on Observer); switched to Party's
process_message_with_checks.

jonas-lj committed:
Address review nits and clean up clippy warnings

- Drop unused R generic parameter from Observer::complete
- Fix two stray blank lines after /// doc comments in Observer
- Replace useless vec![] with array literals in tests

jonas-lj committed:
Move sanity_check_message onto Message

Deduplicates the per-message sanity checks that were copy-pasted
between Party and Observer. The shared method takes nodes, t, and
a closure to derive the encryption random oracle for a given sender.

akichidis commented on fastcrypto-tbls/src/tests/dkg_v1_tests.rs:
```diff
@@ -1017,3 +1018,238 @@ fn test_e2e_dkg_and_key_rotation() {
+
+    // Run DKG as observer:
+
+    // Compute the observers output. The Observer has no shares but only the vss_pk
```
Maybe we should add an assert that indeed no shares produced for Observer

akichidis commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+    /// Compute the output of a DKG protocol given all the dealer messages from the protocol.
+    /// If successful, return the Output which contains the verifying key. The shares field will
+    /// be None since the observer does not hold any shares.
+    pub fn observe_dkg(
```
Do I understand correctly that this is mostly a helper method? Assuming I have all the messages I can just call this to compute the DKG output? Otherwise I have to do the `process_message` my own for one by one the messages and then eventually call the `complete` method?

jonas-lj replied:
Yes, exactly. I think I worked with an assumption that this was exactly what you needed for your integration, but if it's not needed anymore, we could consider getting rid of it.

akichidis commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+        }
+    }
+
+    /// Execute the previous two steps together.
```
I would suggest to mention specifically what the previous steps are specifically. Although someone can tell from the code, but the comment it self depends on some kind of order which is not clear.

akichidis commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+            return Err(FastCryptoError::NotEnoughInputs);
+        }
+
+        info!("DKG: Merging messages with total weight {total_weight}");
```
Is `info` the right log level here?

jonas-lj replied:
`info` is aligned with how it's done for `Party`, but I don't have a strong preference

akichidis replied:
If it's aligned with `Party` then no issue from my side.

akichidis commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+        self.t
+    }
+
+    /// 5. Process a message and create the second message to be broadcasted.
```
I guess for the Observer broadcasting the second message doesn't apply right? Since we just verify the messages here, we do not send any confirmations to other parties

jonas-lj replied:
Yes, you're right - that doc is not accurate. I'll update

akichidis reviewed (commented):
I don't have opinion on the inner crypto mechanics, but thinks look good to me. For sure @benr-ml will need to confirm. 
As an overall comment, and we have discussed this offline, the code duplication with `Party` is not ideal for future maintenance, but if we believe it's safer that way then I am fine with it. Otherwise if there are parts that could safely be deduplicated I would advise to do so.
In any case thanks a lot @jonas-lj for crafting this so quickly 💯

jonas-lj committed:
Polish docs on message types and complete()

Tightens the wording on Complaint, Confirmation, Message,
ProcessedMessage, UsedProcessedMessages, and VerifiedProcessedMessages
to be consistent and to fix a few grammar nits. Names the underlying
steps in complete()'s doc instead of saying "the previous two steps".

jonas-lj committed:
Revert doc polish on Party-only message types

Keeps the Observer::complete doc clarification but restores the
original wording on Complaint, Confirmation, Message, ProcessedMessage,
UsedProcessedMessages, VerifiedProcessedMessages, and Party::complete
to keep this PR focused on the Observer.

jonas-lj committed:
Assert observer outputs have no shares

Pins the invariant that Observer outputs hold only public state, for
both observe_dkg and observe_key_rotation paths.

benr-ml commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+    EG: GroupElement + Serialize + HashToGroupElement + DeserializeOwned,
+    EG::ScalarType: FiatShamirChallenge + Zeroize,
+{
+    /// 1. Create a new ECIES private key and send the public key to all parties.
```
update the comments here and below (e.g., ECIES private key)

benr-ml commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+    /// Alternative to new(), to be used for:
+    /// - Any threshold t, useful in the sync setting.
+    /// - Passing old threshold for key rotation (should be set to None for DKG).
+    pub fn new_advanced(
```
can move to new, no need for new_advanced

benr-ml commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+    }
+
+    /// Alternative to process_message for a known vss.c0 public key, useful for key rotation.
+    pub fn process_message_and_check_pk(
```
no need for this one

benr-ml commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+
+    /// Alternative to complete() - Optimistic version that assumes all messages are valid and if
+    /// other parties received invalid shares, the higher level protocol will handle it.
+    pub fn complete_optimistic(
```
no need

benr-ml commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,421 @@ where
+
+    /// Alternative to complete() - Optimistic variant that interpolates the shares instead of
+    /// summing them, to be used for key rotation. Works only if all weights are 1.
+    pub fn complete_optimistic_key_rotation(
```
no need

jonas-lj replied:
This is because we only need the observer for the DKG and not the key rotation?

benr-ml replied:
right

jonas-lj committed:
Trim Observer to DKG only

Drops observe_key_rotation, complete_optimistic, complete_optimistic_key_rotation,
and process_message_and_check_pk from Observer. observe_dkg now takes
confirmations and delegates to the existing public complete method.

In the e2e test, also verify the post-rotation aggregated signature under
the verifying key that the observer derived during DKG, which exercises
the invariant that key rotation preserves that key.

jonas-lj committed:
Slim Observer API to match streaming callers

Drops Observer::observe_dkg and the unused old_t field/parameter.
Observer::new now takes (nodes, t, random_oracle); merge enforces
total_weight >= t directly. The remaining surface is process_message →
merge → complete, which matches how external callers (e.g. Sui's
randomness manager) drive the DKG.

Test rewritten to use that same streaming pattern.

benr-ml commented on fastcrypto-tbls/src/dkg_v1.rs:
```diff
@@ -983,3 +998,280 @@ where
+        self.t
+    }
+
+    /// 5. Process a message and create the second message to be broadcasted.
```
5?
also numbers below

jonas-lj committed:
Renumber Observer step docs

Observer doesn't have setup steps 1-4 (key gen, node assembly, party
init, create_message), so renumber its own steps 1-4 for clarity.
Also add a doc to Observer::merge (previously undocumented) and drop
the "create the second message to be broadcasted" wording from
process_message — observers don't broadcast anything.
